web/views.py

```diff
@@ -200,18 +208,114 @@ def form_valid(self, form):
         return super().form_valid(form)
 
 
-class StudiesListView(generic.ListView):
+class StudiesListView(generic.ListView, PaginatorMixin, FormView):
     """
     List all active, public studies.
     """
 
+    form_class = forms.StudyListSearchForm
     template_name = "web/studies-list.html"
     model = Study
 
+    def post(self, request, *args, **kwargs):
+        form = self.get_form()
+
+        if form.is_valid():
+            for field in self.form_class().fields:
+                request.session[field] = form.data.get(field, "")
+
+            return self.form_valid(form)
+        else:
+            return self.form_invalid(form)
+
     def get_queryset(self):
-        # TODO if we need to filter by study demographics vs user demographics
-        # or by if they've taken the study before this is the spot
-        return super().get_queryset().filter(state="active", public=True).order_by("?")
+        session = self.request.session
+        user = self.request.user
+        page = self.request.GET.get("page", 1)
+
+        qs = super().get_queryset().filter(state="active", public=True)
+
+        # values from session
+        search_value = session.get("search", "")
+        child_value = session.get("child", "")
+        show_experiments_already_done_value = (
+            session.get("show_experiments_already_done", "") == "on"
+        )
+
+        if search_value:
+            qs = qs.filter(name__icontains=search_value)
+
+        # convert to list as it's no longer being treated as a queryset
+        studies = list(qs)
+
+        if child_value:
+            if user.is_anonymous:
+                # when user is anonymous, child value is an age range.
+                age_range = [int(c) for c in child_value.split(",")]
+                studies = [
+                    s for s in studies if age_range_eligibility_for_study(age_range, s)
+                ]
+            else:
+                # when user is authenticated, child value is a child pk
+                child = Child.objects.get(pk=child_value, user=user)
```

Comment:
this is nice, so far as I can tell this completely takes care of the "pk of a child not associated with this user sent" security issue?

Reply:
It does.  The child object returned must have both the pk of the child and the user object.  The user object is obtained from the the user be authenticated.